src/WebJobs.Script/Description/Node/Compilation/TypeScript/TypeScriptCompilation.cs

```diff
@@ -0,0 +1,134 @@
+        private readonly string _inputFilePath;
+        private readonly TypeScriptCompilationOptions _options;
+        private readonly List<Diagnostic> _diagnostics = new List<Diagnostic>();
+
+        private TypeScriptCompilation(string inputFilePath, TypeScriptCompilationOptions options)
+        {
+            _inputFilePath = inputFilePath;
+            _options = options;
+        }
+
+        public bool SupportsDiagnostics => true;
+
+        private Task CompileAsync()
+        {
+            var tcs = new TaskCompletionSource<object>();
+
+            try
+            {
+                string inputDirectory = Path.GetDirectoryName(_inputFilePath);
+
+                var startInfo = new ProcessStartInfo
+                {
+                    FileName = _options.ToolPath,
+                    RedirectStandardOutput = true,
+                    RedirectStandardError = true,
+                    CreateNoWindow = true,
+                    UseShellExecute = false,
+                    ErrorDialog = false,
+                    WorkingDirectory = inputDirectory,
+                    Arguments = _options.ToArgumentString(_inputFilePath)
+                };
+
+                var process = new Process { StartInfo = startInfo };
+                process.ErrorDataReceived += ProcessDataReceived;
+                process.OutputDataReceived += ProcessDataReceived;
+                process.EnableRaisingEvents = true;
+                process.Exited += (s, e) =>
+                {
+                    process.Close();
+
+                    // TODO: Fix, make sure output is read
+                    Task.Delay(1000).ContinueWith(t =>
+                    {
+                        if (_diagnostics.Any(d => d.Severity == DiagnosticSeverity.Error))
+                        {
+                            tcs.SetException(new CompilationErrorException("Compilation failed.", GetDiagnostics()));
+                        }
+                        else
+                        {
+                            tcs.SetResult(null);
+                        }
+                    });
+                };
+
+                process.Start();
+
+                process.BeginErrorReadLine();
+                process.BeginOutputReadLine();
+            }
+            catch (Exception exc)
+            {
+                tcs.SetException(exc);
+            }
+
+            return tcs.Task;
+        }
+
+        private void ProcessDataReceived(object sender, DataReceivedEventArgs e)
+        {
+            if (e.Data == null)
+            {
+                return;
+            }
+
+            Match match = DiagnosticRegex.Match(e.Data);
+
+            if (match.Success)
+            {
+                DiagnosticSeverity severity = (DiagnosticSeverity)Enum.Parse(typeof(DiagnosticSeverity), match.Groups[4].Value, true);
```

Comment:
Also, what if the version of the compiler changes and the output parsing you're doing fails to parse? How resilient are we?